### ISSUE TITLE
Multi-line records are not supported! panic: pq: unterminated CSV quoted field

### ISSUE THREAD

**Issue by binakot** [timescale/timescaledb-parallel-copy#50]
Got an error on using `timescaledb-parallel-copy`.

```
$ timescaledb-parallel-copy -version
timescaledb-parallel-copy 0.3.0 (linux amd64)

$ timescaledb-parallel-copy --db-name postgres --table table --file data.csv
panic: pq: unterminated CSV quoted field

goroutine 21 [running]:
main.processBatches(0xc0000b6370, 0xc0000926c0)
        /go/src/github.com/timescale/timescaledb-parallel-copy/cmd/timescaledb-parallel-copy/main.go:262 +0x8ad
created by main.main
        /go/src/github.com/timescale/timescaledb-parallel-copy/cmd/timescaledb-parallel-copy/main.go:148 +0x1d2
```

Later I will compile the app from source and try to find the reason and prepare the test data set for reproduction.
The error may be related to the following issues: #19 #24 #31

**Comment by binakot** [comment]
First of all `timescaledb-parallel-copy` gives wrong output of inserted rows when CSV has multiline records. 
Related to issue #19. 

But looks like it works anyway even with wrong result output.

### How to reproduce

Table to insert into:

```sql
CREATE TABLE test (
  id SERIAL PRIMARY KEY,
  state TEXT,
  time TIMESTAMPTZ,
  date TEXT
)
```

Content of test file `data.csv`:

```csv
1,failed,2019-07-29 07:43:14.197455,"---
:image:
  :name: ubuntu:16.04
:before_script:
- echo ""Install sofrware""
- apt-get update && apt-get install -y sudo wget
:script:
- errcount=0
"
2,failed,2019-08-29 08:38:31.747244,"---
:image:
  :name: ubuntu:16.04
:before_script:
- echo ""Install sofrware""
- apt-get update && apt-get install -y sudo wget
:script:
- echo ""Tests""
- ""./tests.sh""
"
3,failed,2019-09-29 08:52:44.77306,"---
:image:
  :name: docker:stable
:services:
- :name: docker:stable-dind
:before_script:
- docker info
:script:
- echo ""Tests""
- ""./tests.sh""
"
4,failed,2019-10-29 09:31:14.098283,"---
:image:
  :name: ubuntu:16.04
:services:
- :name: docker:stable-dind
:before_script:
- docker info
:script:
- echo ""Tests""
- ""./tests.sh""
"
5,failed,2019-11-29 09:55:21.588764,"---
:image:
  :name: docker:stable
:services:
- :name: docker:stable-dind
:before_script:
- docker info
:script:
- echo ""Tests""
- ""./tests.sh""
"
```

```bash
$ timescaledb-parallel-copy --table test --file data.csv
COPY 52
```

52 rows were inserted. 
It's not a count of records in CSV, it is a number or rows in file. 

But table looks correct:

```sql
postgres=# select * from test limit 1;
 id | state  |             time              |                       date                       
----+--------+-------------------------------+--------------------------------------------------
  1 | failed | 2019-07-29 07:43:14.197455+00 | ---                                             
    |        |                               | :image:                                         
    |        |                               |   :name: ubuntu:16.04                           
    |        |                               | :before_script:                                 
    |        |                               | - echo "Install sofrware"                       
    |        |                               | - apt-get update && apt-get install -y sudo wget
    |        |                               | :script:                                        
    |        |                               | - errcount=0                                    
    |        |                               | 
  2 | failed | 2019-08-29 08:38:31.747244+00 | ---                                             
    |        |                               | :image:                                         
    |        |                               |   :name: ubuntu:16.04                           
    |        |                               | :before_script:                                 
    |        |                               | - echo "Install sofrware"                       
    |        |                               | - apt-get update && apt-get install -y sudo wget
    |        |                               | :script:                                        
    |        |                               | - echo "Tests"                                  
    |        |                               | - "./tests.sh"                                  
    |        |                               | 
  3 | failed | 2019-09-29 08:52:44.77306+00  | ---                                             
    |        |                               | :image:                                         
    |        |                               |   :name: docker:stable                          
    |        |                               | :services:                                      
    |        |                               | - :name: docker:stable-dind                     
    |        |                               | :before_script:                                 
    |        |                               | - docker info                                   
    |        |                               | :script:                                        
    |        |                               | - echo "Tests"                                  
    |        |                               | - "./tests.sh"                                  
    |        |                               | 
  4 | failed | 2019-10-29 09:31:14.098283+00 | ---                        
    |        |                               | :image:                    
    |        |                               |   :name: ubuntu:16.04      
    |        |                               | :services:                 
    |        |                               | - :name: docker:stable-dind
    |        |                               | :before_script:            
    |        |                               | - docker info              
    |        |                               | :script:                   
    |        |                               | - echo "Tests"             
    |        |                               | - "./tests.sh"             
    |        |                               | 
  5 | failed | 2019-11-29 09:55:21.588764+00 | ---                        
    |        |                               | :image:                    
    |        |                               |   :name: docker:stable     
    |        |                               | :services:                 
    |        |                               | - :name: docker:stable-dind
    |        |                               | :before_script:            
    |        |                               | - docker info              
    |        |                               | :script:                   
    |        |                               | - echo "Tests"             
    |        |                               | - "./tests.sh"             
    |        |                               | 

(5 rows)
```

**Comment by binakot** [comment]
I seem to have found the cause of the error. It's all about the same problem of multi-line records. 

Specifying the `--batch-size`, which is 5000 by default, is not the number of records, but the number of rows, without taking the fact that the data can be on several lines.

## How to reproduce

Table to insert into:

```sql
CREATE TABLE test (
  id SERIAL PRIMARY KEY,
  state TEXT,
  time TIMESTAMPTZ,
  date TEXT
)
```

Content of test file `data.csv`:

```csv
1,failed,2019-07-29 07:43:14.197455,"---
:image:
  :name: ubuntu:16.04
:before_script:
- echo ""Install sofrware""
- apt-get update && apt-get install -y sudo wget
:script:
- errcount=0
"
```

## Single worker

### --batch-size 1 --workers 1

```bash
$ timescaledb-parallel-copy --table test --file data.csv --batch-size 1 --workers 1

>>> batch=&{[1,failed,2019-07-29 07:43:14.197455,"---]}
>>> copyCmd=COPY "public"."test" FROM STDIN WITH DELIMITER ',' CSV
panic: pq: unterminated CSV quoted field

goroutine 19 [running]:
main.processBatches(0xc0000c4350, 0xc0000ca660)
        /home/binakot/repos/github/timescaledb-parallel-copy/cmd/timescaledb-parallel-copy/main.go:263 +0x978
created by main.main
        /home/binakot/repos/github/timescaledb-parallel-copy/cmd/timescaledb-parallel-copy/main.go:148 +0x1bb

```

### --batch-size 2 --workers 1

```bash
$ timescaledb-parallel-copy --table test --file data.csv --batch-size 2 --workers 1

>>> batch=&{[1,failed,2019-07-29 07:43:14.197455,"--- :image:]}
>>> copyCmd=COPY "public"."test" FROM STDIN WITH DELIMITER ',' CSV
panic: pq: unterminated CSV quoted field

goroutine 6 [running]:
main.processBatches(0xc00001a460, 0xc00002a720)
        /home/binakot/repos/github/timescaledb-parallel-copy/cmd/timescaledb-parallel-copy/main.go:263 +0x978
created by main.main
        /home/binakot/repos/github/timescaledb-parallel-copy/cmd/timescaledb-parallel-copy/main.go:148 +0x1bb
```

### --batch-size 3 --workers 1

```bash
$ timescaledb-parallel-copy --table test --file data.csv --batch-size 3 --workers 1

>>> batch=&{[1,failed,2019-07-29 07:43:14.197455,"--- :image:   :name: ubuntu:16.04]}
>>> copyCmd=COPY "public"."test" FROM STDIN WITH DELIMITER ',' CSV
panic: pq: unterminated CSV quoted field

goroutine 6 [running]:
main.processBatches(0xc00001a460, 0xc00002a720)
        /home/binakot/repos/github/timescaledb-parallel-copy/cmd/timescaledb-parallel-copy/main.go:263 +0x978
created by main.main
        /home/binakot/repos/github/timescaledb-parallel-copy/cmd/timescaledb-parallel-copy/main.go:148 +0x1bb
```

### --batch-size 100 --workers 1

```bash
$ timescaledb-parallel-copy --table test --file data.csv --batch-size 100 --workers 1

COPY 9
```

## Two workers

### --batch-size 1 --workers 2

```bash
$ timescaledb-parallel-copy --table test --file data.csv --batch-size 1 --workers 2

>>> batch=&{[:image:]}
>>> copyCmd=COPY "public"."test" FROM STDIN WITH DELIMITER ',' CSV
panic: pq: invalid input syntax for integer: ":image:"

goroutine 19 [running]:
main.processBatches(0xc0000c4350, 0xc0000ca660)
        /home/binakot/repos/github/timescaledb-parallel-copy/cmd/timescaledb-parallel-copy/main.go:263 +0x978
created by main.main
        /home/binakot/repos/github/timescaledb-parallel-copy/cmd/timescaledb-parallel-copy/main.go:148 +0x1bb
```

---

Debug output done with code `fmt.Printf(">>> batch=%s\r\n>>> copyCmd=%s\r\n", batch, copyCmd)` here: https://github.com/timescale/timescaledb-parallel-copy/blob/master/cmd/timescaledb-parallel-copy/main.go#L262

The current implementation is completely dependent on reading the file line by line. We need to think about how to ensure that the data is read by records rather than rows. Because every worker can get to the middle of a multi-line record.

**Comment by binakot** [comment]
@RobAtticus Hello!

Any ideas how to fix that? 
May be is there some hacks to fix multi-line records? 
Or is it required additional implementation?

**Comment by ryanmjacobs** [comment]
Also had this issue. Had to resort to `psql -d new_db -c "\COPY conditions FROM old_db.csv CSV"`

**Comment by mindrunner** [comment]
I am facing an t least similar issue. Not entirely sure if that's related. Quick introduction about my setup:

- We have a `mysql` database with a rest-api to access its data. From there, I do import several records into the `timescaledb`. One of the tables contains a `json` column with `PEM` encoded certificates (including the usual linebreaks in base64). On timescaledb/postgres side, I store those certificates / keys in separate columns (`text`) (no more `JSON` column)

- Using `psql` to create backups / dumps of the entire timescale database.

- While testing the backups by restoring locally I encountered an error (only on my production data, test instance was fine):

```
panic: ERROR: missing data for column "emailAddress" (SQLSTATE 22P04)

goroutine 6 [running]:
main.processBatches(0x0?, 0x0?)
	/home/le/go/pkg/mod/github.com/timescale/timescaledb-parallel-copy@v0.4.0/cmd/timescaledb-parallel-copy/main.go:220 +0x665
created by main.main
	/home/le/go/pkg/mod/github.com/timescale/timescaledb-parallel-copy@v0.4.0/cmd/timescaledb-parallel-copy/main.go:145 +0x4ad
COPY 0, took 15.694658ms with 16 worker(s) (mean rate 0.000000/sec)
```

`emailAddress` is the second column in that table, after primary key (`id`), so that was little confusing, but I suspected the multiline certs. Turnes out, the `production` database containing some certs with `\r\n` linebreaks, whereas on `test`, there are only `\n` ones. Seems like `timescaledb-parallel-copy` does not like the `\r\n` variant (at least in my case). 

I haven't fixed that up yet in order to run another test, so for now all this is pure assumptions. (looks pretty straight forward, tho).
